src/speculators/data_generation/vocab_mapping.py

```diff
@@ -0,0 +1,76 @@
+"""Vocabulary mapping utilities for draft model training."""
+
+import os
+from collections import Counter
+
+import torch
+from datasets import Dataset as HFDataset
+from tqdm import tqdm  # type: ignore[import-untyped]
+
+__all__ = [
+    "build_vocab_mappings_from_distribution",
+    "save_token_frequency_distribution",
+]
+
+
+def save_token_frequency_distribution(
+    dataset: HFDataset,
+    cache_dir: str = "./cache/token_frequencies",
+    cache_key: str = "token_freq",
+) -> str:
+    """Save token frequency distribution from the dataset."""
+    os.makedirs(cache_dir, exist_ok=True)
+    freq_dist_path = os.path.join(cache_dir, f"{cache_key}_token_freq.pt")
+
+    if os.path.exists(freq_dist_path):
+        return freq_dist_path
+
+    token_freq: Counter[int] = Counter()
+    for item in tqdm(dataset, desc="Counting token frequencies"):
+        input_ids = item["input_ids"]
+        loss_mask = item["loss_mask"]
+        masked_ids = input_ids[loss_mask == 1]
+        unique_ids, counts = masked_ids.unique(return_counts=True)
+        batch_token_freq = dict(zip(unique_ids.tolist(), counts.tolist(), strict=False))
```

Comment:
Should use `strict=True`. If these don't line up then it means we have a different number of unique_ids than counts for those ids which would be a problem.